File: backend/app/auth/roles.py
"""RBAC roles, ordered by privilege. Higher index = more privilege."""
from __future__ import annotations

ROLES = ["viewer", "operator", "deployer", "admin", "owner"]
_RANK = {r: i for i, r in enumerate(ROLES)}


def rank(role: str) -> int:
    return _RANK.get(role, -1)


def has_at_least(role: str, minimum: str) -> bool:
    return rank(role) >= rank(minimum)


# What each role can do (documentation + coarse gating reference):
#   viewer   — read everything
#   operator — + ack alerts, manual heal, restart containers
#   deployer — + deployments, rollback, healing policies
#   admin    — + users, API keys, dashboards regen, alert rules
#   owner    — + tenant management

File: frontend/src/auth/roles.ts
// Mirrors the backend RBAC role order (app/auth/roles.py).
export const ROLES = ["viewer", "operator", "deployer", "admin", "owner"] as const;

export function roleAtLeast(role: string, minimum: string): boolean {
  return ROLES.indexOf(role as never) >= ROLES.indexOf(minimum as never);
}
